### DISCUSSION TITLE
I was looking at Content Security Policy from the forum here and got lost. If anyone can clarify.

### DISCUSSION THREAD

**Discussion by Inzendis** [vercel/next.js#58521] [App Router]
Im on Next JS 13.5.6

So I was trying to do a CSP and its blocking some self scripts. So meta tags that I put myself are being blocked. 

Content Security Policy blocks inline execution of scripts and stylesheets
The Content Security Policy (CSP) prevents cross-site scripting attacks by blocking inline execution of scripts and style sheets. So everything that is a meta tag alongside stuff from _next, node_modules and internal.js are being blocked. 

How do I enable everything in my site PWA? Im not using cdns or using something from an external link. 

I was looking at nonces, but I can't seem to apply the nonce to all the things that appear as script. I can apply the nonce to my meta tags, but thats about it? 

How one of the short ones look like from console:

`self.__next_f.push([1,"1:HL[\"/_next/static/css/app/layout.css?v=1700106217959\",\"style\"]\n0:\"$L2\"\n"])`

![image](https://github.com/vercel/next.js/assets/80369840/bc36d5ec-c8a0-4085-8c24-a549f3707798)

I'm new to this while CSP thing, any help would be apreciated.

In my middleware
```
const cspHeader = `
    frame-ancestors 'none';
    default-src 'self';
  `
      .replace(/\n/g, "")
      .trim();
    const response = NextResponse.next();
    response.headers.set("Content-Security-Policy", cspHeader);
    response.headers.set("X-Frame-Options", "DENY");
    response.headers.set("X-Content-Type-Options", "nosniff");
```

Edit: fixed it slightly as you can see

![image](https://github.com/vercel/next.js/assets/80369840/1e772168-515f-40e0-a7bf-6438138dc17b)

This the only one missing to fix, this one idk where it comes from

![image](https://github.com/vercel/next.js/assets/80369840/394d7529-aa3f-4884-9b28-dfe5de6ec529)

**Comment by 0xdbe** [comment]
Hi @Inzendis

You have to know that `eval()` is used by Webpack and react-refresh for LiveReload. This feature is used only in local development environment.

This can be easily fix by adding `unsafe-eval` for dev env only:

```
const isDev = process.env.NODE_ENV === 'development' ? true : false

const contentSecurityPolicyDirectiveScriptSrcValues = isDev
  ? [`'unsafe-eval'`]
  : // script-src directives for prod 
```
